### ISSUE TITLE
Wrong implementation of method log_prob in torch.distributions.negative_binomial

### ISSUE THREAD

**Issue by schmnin** [pytorch/pytorch#62178]
## 🐛 Bug

The method log_prob is wrongly implemented in the negative binomial distribution if one assumes that the attribute probs is the event probability of failures as stated in the documentary.
In the code it is assumed that 
log_unnormalized_prob = (self.total_count * F.logsigmoid(-self.logits) + value * F.logsigmoid(self.logits))

According to the definition of the negative binomial distribution this is supposed to be equal to 
self.total_count * log(probs) + value * log(1-probs) 
(see e.g. https://en.wikipedia.org/wiki/Negative_binomial_distribution).

However: 
logsigmoid(logits) = log(1/(1+exp(-logits))) = log(1/(1+exp(-log(p/(1-p))))) = log(1/(1+exp(log((1-p)/p))) = log(1/(1+(1-p)/p)) = log(1/(1/p)) = log(p) 

## Additional context
Maybe this happened due to https://github.com/pytorch/pytorch/issues/42449

**Comment by fritzo** [comment]
cc @jamestwebber

**Comment by jamestwebber** [comment]
Thanks for finding this @schmnin! If I'm reading this right, the error is that `logits` and `-logits` need to swap places? e.g. it should be 

```
log_unnormalized_prob = (self.total_count * F.logsigmoid(self.logits) + value * F.logsigmoid(-self.logits))
```

**Comment by schmnin** [comment]
Yes, that would be one way to solve the issue of the log_prob method. Alternatively you can change the documentation of probs back to "success probability".  I did not check the other methods and only had a short glampse once more in the code, but I think the mean of the negative binomial expects p to be the success probability, too. Maybe in the end, the change from "success probability" to "failure probabilty" in the documentation was wrong or some changes happened to both the documentation and the code?

**Comment by jamestwebber** [comment]
Wait 🤦 is this even wrong? In the pytorch implementation, `logits` is the log-odds of _failure_, whereas the equation quoted from Wikipedia (which lists a bunch of different PMFs) is likely in terms of success probabilities. That would explain a swap in the signs.

Clearly the current implementation is confusing to everyone, although I would posit that the Wikipedia article is also confusing and maybe we should edit that too. 🙂

**Comment by schmnin** [comment]
In the equations I wrote above I used the same meaning of p, both being the probability of failures (obviously I therefore used 1-p_failure wherever wikipedia writes p_success). 
To illustrate shortly why something definitely is fishy about the implementations: So wikipedia states the mean as p_success*total_count / (1-p_success), but currently in pytorch it is calculated as total_count * exp(logits(p_failure)) = total_count* p_failure/(1-p_failure). 

But yes, I totally agree that the definitions of the negative binomial are very confusing, especially since there exist 3 different names for all of the parameters 🤦

**Comment by jamestwebber** [comment]
> To illustrate shortly why something definitely is fishy about the implementations: So wikipedia states the mean as p_success * total_count / (1-p_success), but currently in pytorch it is calculated as total_count * exp(logits(p_failure)) = total_count * p_failure/(1-p_failure).

Just recreating the experiments in #42449, it seems like the mean calculation is correct. I think this is just confusing.

**Comment by jamestwebber** [comment]
```
from scipy.stats import nbinom
from torch.distributions import NegativeBinomial
​
n = 2.0
x = 0.0
p = 0.75
​
NegativeBinomial(n, 1 - p).log_prob(torch.tensor(x)).exp(), nbinom.pmf(x, n, p)
# out: (tensor(0.5625), 0.5625)

NegativeBinomial(n, p).log_prob(torch.tensor(x)).exp(), nbinom.pmf(x, n, 1 - p)
# out: (tensor(0.0625), 0.0625)

NegativeBinomial(n, 1 - p).mean, nbinom.mean(n, p)
# out: (tensor(0.6667), 0.6666666666666665)

NegativeBinomial(n, p).mean, nbinom.mean(n, 1 - p)
# out: (tensor(6.), 6.0)
```

**Comment by jamestwebber** [comment]
> wikipedia states the mean as p_success * total_count / (1-p_success)

Ah, _this_ is wrong...as p_success approaches zero the mean should go to infinity, but in this equation it would go to 0.

**Comment by schmnin** [comment]
okay, so n is the number of successes. but the pytorch total count is actually the number of failures, isn't it? I have to admit, that I also struggled with the pytorch documentation there: 
"non-negative number of negative Bernoulli
trials to stop, although the distribution is still valid for real
valued count" (pytorch docu) sounds to me like "number of failures until the experiment is stopped" (wiki) and not like " n is the number of successes" (scipy)
So you can't use n in the NegativeBinomial class of pytorch directly, I think.

**Comment by SplitInfinity** [comment]
It's been a while since I looked at this but I think that thanks to #42449, the documentation is correct but the API is not intuitive and doesn't match `scipy/numpy`. We outlined steps to fix this in #42449 (but never took them haha):

https://github.com/pytorch/pytorch/issues/42449#issuecomment-702444295

Is that accurate?

**Comment by lcontento** [comment]
If I understood correctly, the OP in issue https://github.com/pytorch/pytorch/issues/42449 claimed that the fact that `torch.distributions.NegativeBinomial(n, p)` and `scipy.stats.nbinom(n, 1-p)` were the same distribution was in contradiction with the documentation at the time (`probs` as success probability). I don't believe there was any contradiction in the first place, so that PR https://github.com/pytorch/pytorch/pull/45693 was not warranted.

A sketch of my reasoning:
```
scipy.stats.nbinom(n, p)
```
This is the distribution of the number of failed iid Bernoulli trials before `n` successes are achieved (`p` is the success probability). See https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.nbinom.html

```
torch.distributions.NegativeBinomial(n, p)
```
According to the original documentation, this is the distribution of the number of successful iid Bernoulli trials before `n` failures are achieved (`p` is the success probability).
By calling success what was a failure, this is equivalent to: distribution of the number of failed iid Bernoulli trials before `n` successes are achieved (`1 - p` is the success probability).
And this is `scipy.stats.nbinom(n, 1-p)`, so no contradiction ever happened.

I think the OP of https://github.com/pytorch/pytorch/issues/42449 got confused on the meaning of the `n`, which is failures in the `pytorch` case and successes for `scipy`. And indeed all these competing definitions for the negative binomial make things quite confusing, so by all means someone please double check my reasoning above ;-)

**Comment by jamestwebber** [comment]
That makes sense, I think that's what happened.

I'm not sure what the best way forward is...it seems like getting from the current code to something that matches scipy would require a few cycles of deprecation (`total_count` -> `num_failures`, and then switch to using `num_successes` eventually).

Luckily that's not my call to make :joy:

**Comment by malfet** [comment]
@albanD says that definition is correct, we have to be careful in documenting the difference between calculating probability of success vs probability of failure

**Comment by GeoffNN** [comment]
Hi

I'm confused by the documentation here.
What exactly is the PMF of the implemented function ?

From the docstring, I thought it was

```python

    log_prob = (
        torch.lgamma(value + self.total_count)  # (value+self.total_count-1)!
        - torch.lgamma(total_count)  # /((self.total_count-1)!)
        - torch.lgamma(
            value + 1
        )  # /(value!)
        + self.total_count * torch.log1p(-self.probs)  # (1-p)^r
        + value * torch.log(self.probs)  # p^d
    )
```

But this actually gives a different result than the implemented log_probs function.

**Comment by jamestwebber** [comment]
> Is there an advantage of using the logits rather than the probs in this implementation? Using the logits is more confusing/harder to follow.

logits are more useful when doing backprop through the distribution, because there's a lot more room for very small and very high probabilities. If an optimizer pushes the probs to 0 or 1 it will start to generate NaNs if it is fed any incompatible data.
